[New Feature] Redis sessions handler

alikon opened the pull request:
This pr add a Redis session handler
Selectable from Global configuration -> Session settings

![redis_session_settings](https://cloud.githubusercontent.com/assets/181681/9150471/634f0bf2-3dda-11e5-8a72-ab91f6216652.png)
#### Test environment

Redis server. installed and setted  http://redis.io/topics/quickstart
Redis PHP driver pecl installed https://github.com/phpredis/phpredis#installation
Redis webgui  http://joeferner.github.io/redis-commander/ (require node,js)
#### Expected result

All works as expected with some performance gain
login & logout 
session expiration
modules: whoisonline, logged, stats
When the Redis handler is selected the #__session table  is not used

alikon committed:
redis partone

alikon committed:
redis partone fix

alikon committed:
redis session user plugin

alikon committed:
redis part2

alikon committed:
redis part 3

alikon committed:
redis fix2

alikon committed:
redis fix4

alikon committed:
fix commit

dgt41 commented:
Thanks Nicola for this one, the ability to use redis for the sessions is a welcome feature.
The performance increase is obvious but still haven’t done some benchmarking.

I do get some notices tho:
![screenshot 2015-08-09 20 01 20](https://cloud.githubusercontent.com/assets/3889375/9155935/00581478-3ed2-11e5-907f-00e452611fa0.png)

@mbabker @wilsonge This one is touching framework files factory and sessions.

alikon commented:
@dgt41 for avoid the mod_logged notice you should use the alternative module layout redis

i wish to listen a lot of comments,  cause my code is not OO i'm afraid

zero-24 commented:
@alikon 

I have just opend a new PR against yours to fix CS issues and do some simplifications ;) As well es public / private changes.

Can you have a look / review here: https://github.com/alikon/joomla-cms/pull/9

>@dgt41 for avoid the mod_logged notice you should use the alternative module layout redis

Fixed with: https://github.com/alikon/joomla-cms/pull/9

There are also 2 or 3 `@todo` marks can you commet there so we can fix it?

zero-24 commented:
@alikon 

>@dgt41 for avoid the mod_logged notice you should use the alternative module layout redis

There is no alternative module layout redis for mod_logged. Only for mod_whoisonline (issue gets fixed with https://github.com/alikon/joomla-cms/pull/9)
Can you explain the changes that we need to do for the mod_logged layout?

zero-24 commented:
Just 2 things i don't fix. We have 2 new language strings the needs to be added to the language files:

```
JERROR_SESSION_REDIS_DESTROY
JERROR_SESSION_REDIS_READ
```

waader commented:
I assume that getListFromFromRedis should be named getListFromRedis, right?

zero-24 commented:
Yes. :smile:  @alikon @waader fixed with: https://github.com/zero-24/joomla-cms/commit/53bfa247207291a315f810f517848cd7f590d502

alikon committed:
Zero24+  commit

alikon commented:
i've confused the modules, sorry... 
@zero-24  thx 1000 but i've made some others dirty bit  :smile:

alikon committed:
17765.1
FILE: /home/travis/build/joomla/joomla-cms/modules/mod_whosonline/helper.php

alikon committed:
travis

alikon committed:
joomla user/joomla/joomla.php travis fix

alikon committed:
libraries/joomla/factory.php travis fix

alikon committed:
plugins/user/joomla/joomla.php travis cs

alikon committed:
libraries/joomla/session/storage/redis.php travis cs fix

alikon committed:
libraries/joomla/session/redis.php  travis cs fix

alikon committed:
administrator/modules/mod_logged/helper.php travis fix

zero-24 committed:
Update helper.php

zero-24 committed:
Update cms.php

zero-24 committed:
Update redis.php

zero-24 committed:
Update joomla.php

zero-24 commented:
@alikon Hupfully the last PR series makes travis finaly happy ;)

alikon committed:
Merge pull request #11 from zero-24/patch-2

Fix Travis for mod_status/helper.php

alikon committed:
Merge pull request #12 from zero-24/patch-6

Fix Travis for: libraries/cms/application/cms.php

alikon committed:
Merge pull request #13 from zero-24/patch-9

Fix Travis for: /libraries/joomla/session/storage/redis.php

alikon committed:
Merge pull request #14 from zero-24/patch-10

Fix Travis for: plugins/user/joomla/joomla.php

waader commented:
Redis as a session handler can be selected even if it is not installed/present. So when you do that, after saving you get this message: "Fatal error: Class 'Redis' not found in joomla34\libraries\joomla\factory.php on line 846"

Ideally redis should not be an option to choose from if it is not installed.

mbabker commented:
The Redis session storage class needs an `isSupported` method (like [the one for Memcache](https://github.com/joomla/joomla-cms/blob/3.4.3/libraries/joomla/session/storage/memcache.php#L70-L80)) to do this check.  Otherwise the base class it extends from always returns true.

zero-24 committed:
Update cms.php

zero-24 committed:
Update redis.php

zero-24 committed:
Update joomla.php

zero-24 commented on administrator/modules/mod_status/helper.php:
```diff
@@ -0,0 +1,137 @@
+				$data = array();
+			}
+
+			// @todo: If the try fails $data is not defined here.
```
this todo can be removed. as it get fixed with https://github.com/joomla/joomla-cms/pull/7659/files#diff-ae5f3a0d8be5b3afb8c40dd2dfd55718R89

waader commented:
When changing the redis port there should be a check if the server is responding. Otherwise one gets this error message "Error displaying the error page: Application Instantiation Error: Error pinging REDIS database" after saving the settings.

I also had a look in configuration.php and there I found these settings:
public $redis_server_port = '6379'
public $session_server_port = '6378'

The port change is no reflected in $redis_server_port. Also I am wondering why $redis_server_port is needed at all?

alikon committed:
fix travis

 866 | ERROR | Whitespace found at end of line

alikon committed:
Merge pull request #15 from zero-24/patch-2

Travis know CS better than me :(

alikon committed:
added isSupported()

added isSupported

alikon commented:
Redis selectable only when available should be fixed now

alikon commented:
For the check of connection  
What should be the best approach ?
An ajax check triggered on change (port)....
The same check should be triggered on the db settings tab too ,cause if you change for example the dB nsme no check is triggered and you get an application error similar to the redis wrong port

waader commented:
@alikon In your code you have variable names like utenti or elm. Can you change that to somewhat more descriptive.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="http://issues.joomla.org/tracker/joomla-cms/7659">issues.joomla.org/joomla-cms/7659</a>.</sub>

alikon commented:
I'm planning to rewrite this pr a little bit to add a check  on connection settings  as per your previous comment and surely I'll translate to English the Italian variable names. Thx for all your precious  comments

alikon committed:
redis partwo

added language string for connection check failed

alikon committed:
redis partwo

class rename

alikon committed:
redis partwo

class rename

alikon committed:
redis partwo

class rename

alikon committed:
redis partwo

class rename

alikon committed:
redis partwo

class rename

alikon committed:
redis partwo

class rename

alikon committed:
redis partwo

class rename

alikon committed:
redis partwo

manage cache and session

alikon committed:
redis partwo

redis class as key-value datastore

alikon committed:
redis partwo

added connection check for redis when selected as session handler e/o cache handler

alikon committed:
redis partwo - travis fix

travis fix

alikon committed:
redis partwo - travis fix

travis fix

alikon committed:
redis partwo - travis fix

travis fix

alikon committed:
redis partwo - travis fix

travis fix

alikon committed:
redis partwo - travis fix

travis fix

alikon committed:
redis partwo - travis fix

travis fix

alikon committed:
redis partwo - travis fix

travis fix

alikon committed:
fix check connection

fixed the check of connectable cache connetction handler

brianteeman commented:
@alikon is this still on your todo list or has it been abandoned and should be closed?<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/7659">issues.joomla.org/joomla-cms/7659</a>.</sub>

alikon commented:
absolutely yes, i hope with the help of a joomla gsoc 2016 student this task will be done, even better, 
https://docs.joomla.org/GSOC_2016_Project_Ideas#Project_V:_NoSQL_Database_Driver_.28Redis.2C_MongoDB.29 , finger crossed

mbabker commented:
Can I suggest this be closed pending the results of GSoC?  At a quick glance at the current state of this PR, there are too many moving parts and too many possible B/C breaks for this implementation to be successful (for example, the changes to how the users are always stored in the database even if the database isn't the session store removes this tracking completely for any handler that isn't the database, "none" (filesystem), or Redis).

brianteeman commented:
You can suggest anything :D

brianteeman commented:
Based on the comment above I am closing this at this moment in time<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/7659">issues.joomla.org/joomla-cms/7659</a>.</sub>
